Probability of continuous random variables lying in a finite set is now zero

akash9712 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->


#### Brief description of what is fixed or changed
Probability of any continuous random variable with a finite density lying
on a finite number of points is zero. Also, probability when the CRVs are not equal to a given point is one. Added tests for the same.
    
 Example(before fix):
 ```
 >>> E = Exponential('E', 3)
 >>> P(Eq(E, 1))
 ...
 AttributeError: 'FiniteSet' object has no attribute 'as_coeff_Mul'
>>> P(Ne(E, 1)) 
```
  
 Example(after fix):
 ```
 >>> E = Exponential('E', 3)
 >>> P(Eq(E, 1))
 0
>>> P(Ne(E, 1))
1
 ```


#### Other comments

akash9712 committed:
Probability of continuous random variables lying in a finite set is now zero

Probability of any continuous random variable with a finite density lying
on a finite number of points is zero.

Example(before fix):
```
>>> E = Exponential('E', 3)
>>> P(Eq(E, 1))
...
AttributeError: 'FiniteSet' object has no attribute 'as_coeff_Mul'
```

Example(after fix):
```
>>> E = Exponential('E', 3)
>>> P(Eq(E, 1))
0
```

akash9712 committed:
Added newline at the end of test_continuous_rv.py

normalhuman commented on sympy/stats/crv.py:
```diff
@@ -340,10 +340,11 @@ def probability(self, condition, **kwargs):
             pdf = self.compute_density(rv, **kwargs)
             # return S.Zero if `domain` is empty set
-            if domain.set is S.EmptySet:
+            domain_set = domain.set
```
Why introduce a variable for this? domain.set does not do anything time-costly, it just returns `domain.args[1]`. There shouldn't be any noticeable overhead of using domain.set three times in this code, especially compared to Integral below.

akash9712 replied:
I introduced the variable because I thought that instead of calling the method thrice, it would be better to store the value. You are right, compared to the integral, the time is negligible. I have updated the PR accordingly.

normalhuman reviewed (commented):
Looks good. 

Ideas which are perhaps better suited for separate projects: `P(Ne(X, 4))` and `P(Eq(X, 5), Eq(X, 4))` both throw errors when X is a continuous RV.

akash9712 committed:
Removed extra variable from `probability` in crv.py

akash9712 commented:
@normalhuman `P(Ne(X, 4))` can be fixed in the same function. Should I push the changes in this PR?

normalhuman commented:
If the solution for Ne is similar, it makes sense for it to be in this PR.

akash9712 committed:
Probability of continuous RV not lying on a given point is now one

Probability of any continuous random variable with a finite density given
to be not equal to a given point returns one. Added tests.

akash9712 commented:
@normalhuman There's a problem with the current solution for `Ne`. In case the given point is a symbol,  the expression returned can't be simplified further, and thus, does not evaluate to one. For example:
```
>>> N = Normal('N', 2, 3)
>>> x = Symbol('x')
>>> P(Ne(N, x))
-5*sqrt(2)*(-sqrt(2)*pi*(-x + 3)*exp((-x + 3)**2/50)*erf(sqrt(2)*(-x + 3)/10)/10 - sqrt(2)*pi*(-x + 3)*exp((-x + 3)**2/50)/10)*exp(-x**2/50 + 3*x/25 - 9/50)/(2*pi*(-x + 3)) + 5*sqrt(2)*(-sqrt(2)*pi*(-x + 3)*exp((-x + 3)**2/50)*erf(sqrt(2)*(-x + 3)/10)/10 + sqrt(2)*pi*(-x + 3)*exp((-x + 3)**2/50)/10)*exp(-x**2/50 + 3*x/25 - 9/50)/(2*pi*(-x + 3))
```

akash9712 committed:
Evaluate probability with `Ne` conditions for continuous RV

Probability conditions with where given condition is an instance of `Ne`
is evaluated as `1 - P(Eq())` for continuous RVs.

akash9712 committed:
Remove check for Union in `probability` in crv.py

akash9712 commented:
Ping @normalhuman

normalhuman commented on sympy/stats/crv.py:
```diff
@@ -16,7 +16,7 @@
 from sympy import (Interval, Intersection, symbols, sympify, Dummy, Mul,
         Integral, And, Or, Piecewise, cacheit, integrate, oo, Lambda,
-        Basic, S, exp, I)
+        Basic, S, exp, I, FiniteSet, Union, Interval, Ne, Eq)
```
Do you still need to add the imports of Union and Interval?

normalhuman commented on sympy/stats/crv.py:
```diff
@@ -355,7 +359,8 @@ def probability(self, condition, **kwargs):
             space = SingleContinuousPSpace(z, dens)
-            return space.probability(condition.__class__(space.value, 0))
+            result = space.probability(condition.__class__(space.value, 0))
+            return result if not cond_inv else 1 - result
```
For consistency with other returns, this should be `S.One - result`, since we are returning a SymPy object. Probably, 1 would be sympified to S.One anyway, but explicit form is better.

akash9712 committed:
Removed `Union` import and used S.One instead of 1 in crv.py

akash9712 commented:
Ping @normalhuman.  Made the changes as suggested,
